op-node/l1/batching.go

```diff
@@ -2,134 +2,181 @@ package l1
+	"github.com/hashicorp/go-multierror"
 )
 
-var (
-	TooManyRetries = errors.New("too many retries")
-)
+// IterativeBatchCall is an util to create a job to fetch many RPC requests in batches,
+// and enable the caller to parallelize easily and safely, handle and re-try errors,
+// and pick a batch size all by simply calling Fetch again and again until it returns io.EOF.
+type IterativeBatchCall[K any, V any, O any] struct {
+	completed uint32       // tracks how far to completing all requests we are
+	resetLock sync.RWMutex // ensures we do not concurrently read (incl. fetch) / reset
 
-// parallelBatchCall creates a drop-in replacement for the standard batchCallContextFn that splits requests into more batch requests, and will parallelize and retry as configured.
-func parallelBatchCall(log log.Logger, getBatch batchCallContextFn, maxRetry int, maxPerBatch int, maxParallel int) batchCallContextFn {
-	return func(ctx context.Context, requests []rpc.BatchElem) error {
-		return fetchBatched(ctx, log, requests, getBatch, maxRetry, maxPerBatch, maxParallel)
-	}
-}
+	requestsKeys []K
+	batchSize    int
 
-type batchResult struct {
-	failed  []rpc.BatchElem // if anything has to be retried
-	err     error           // if the batch as a whole failed
-	success int             // amount of items that completed successfully
-}
+	makeRequest func(K) (V, rpc.BatchElem)
+	makeResults func([]K, []V) (O, error)
+	getBatch    batchCallContextFn
 
-// fetchBatched fetches the given requests in batches of at most maxPerBatch elements, and with at most maxRetry retries per batch.
-// Batch requests may be split into maxParallel go-routines.
-// Retries only apply to individual request errors, not to the outer batch-requests that combine them into batches.
-func fetchBatched(ctx context.Context, log log.Logger, requests []rpc.BatchElem, getBatch batchCallContextFn, maxRetry int, maxPerBatch int, maxParallel int) error {
-	batchRequest := func(ctx context.Context, missing []rpc.BatchElem) (failed []rpc.BatchElem, err error) {
-		if err := getBatch(ctx, missing); err != nil {
-			return nil, fmt.Errorf("failed batch-retrieval: %w", err)
-		}
-		for _, elem := range missing {
-			if elem.Error != nil {
-				log.Trace("batch request element failed", "err", elem.Error, "elem", elem.Args[0])
-				elem.Error = nil // reset, we'll try this element again
-				failed = append(failed, elem)
-				continue
-			}
-		}
-		return failed, nil
-	}
+	requestsValues []V
+	scheduled      chan rpc.BatchElem
 
-	// limit capacity, don't write to underlying array on retries
-	requests = requests[:len(requests):len(requests)]
+	results *O
+}
 
-	expectedBatches := (len(requests) + maxPerBatch - 1) / maxPerBatch
+// NewIterativeBatchCall constructs a batch call, fetching the values with the given keys,
+// and transforms them into a verified final result.
+func NewIterativeBatchCall[K any, V any, O any](
+	requestsKeys []K,
+	makeRequest func(K) (V, rpc.BatchElem),
```

Comment:
`makeRequest` is fine for now but would have to be refactored if we wanted to use it with something that took more than 1 argument in the rpc params

Reply:
`K` can be any type, e.g. a bundle of request arguments, and the `makeRequest` function can then output a batch request with multiple RPC arguments by unpacking it in some request specific way.

Reply:
Ah i wasn't sure if it would need to be `[]K`